File: factory/factory.py
from abc import ABC, abstractmethod


class Animal(ABC):
    @abstractmethod
    def speak(self):
        pass


class Dog(Animal):
    def speak(self):
        return 'Woof!'


class Cat(Animal):
    def speak(self):
        return 'Meow!'


class AnimalFactory:
    @staticmethod
    def create_animal(type):
        if type == 'dog':
            return Dog()
        elif type == 'cat':
            return Cat()
        else:
            raise ValueError('Invalid animal type')


if __name__ == '__main__':
    dog = AnimalFactory.create_animal('dog')
    cat = AnimalFactory.create_animal('cat')

    print(dog.speak()) # Output: Woof!
    print(cat.speak()) # Output: Meow!

File: factory/factory.js
class Animal {
    speak() {
        throw new Error('Method not implemented');
    }
}

class Dog extends Animal {
    speak() {
        return 'Woof!';
    }
}

class Cat extends Animal {
    speak() {
        return 'Meow!';
    }
}

class AnimalFactory {
    static createAnimal(type) {
        if (type === 'dog') {
            return new Dog();
        } else if (type === 'cat') {
            return new Cat();
        } else {
            throw new Error('Invalid animal type');
        }
    }
}

const dog = AnimalFactory.createAnimal('dog');
const cat = AnimalFactory.createAnimal('cat');

console.log(dog.speak());
console.log(cat.speak());
